File: code/ruby-ver/gg.rb
#class which finds the min canonical form for any graph

class Canonical_Form
  def initialize
    @min_cf
  end

  #make a deep copy of the graph and flip it
  def flip_graph g
    fg = g.deep_copy
    fg.adj_list.each{|n|
      n.nbrs.reverse!
      n.edges.reverse!}
    return fg
  end

  #return the min cf using clockwise bfs over the given graph and its flip
  def find_cf g
    #g = g.deep_copy
    f = flip_graph g

    for d in 0..1
      if d == 0
        graph = g
      else
        graph = f
      end

      graph.adj_list.each {|n|
      n.nbrs.each {|nbr|
        bfi = CW_BFS.new(graph,n,graph.adj_list[nbr]).search
        cf = create_graph_from_bfi(graph,bfi)
        if @min_cf.nil?
          @min_cf = cf
        else
          compare_cf(cf)
        end}}
    end
    return @min_cf
  end


  #
  def create_graph_from_bfi(g,bfi)
    cf = g.deep_copy
    cf.adj_list.each{|n|
      n.label = bfi.nodes[n.label]
      n.nbrs.map!{|nbr| bfi.nodes[nbr]}
      n.edges.map!{|e| bfi.edges[e.abs]}
    }

    cf.adj_list.sort_by! {|n| n.label}
    cf.adj_list.each{|n|
      pos = n.nbrs.index(n.nbrs.min)
      n.shift_nbrs pos}
    return cf
  end

  def  compare_cf cf
    @min_cf.adj_list.each_index {|i|
      min_cf_node = @min_cf.adj_list[i]
      cf_node = cf.adj_list[i]

      min_cf_node.nbrs.each_index{|j|
        min_nbr = min_cf_node.nbrs[j]
        nbr = cf_node.nbrs[j]

        if min_nbr > nbr
          @min_cf = cf
        end
        if min_nbr < nbr
          return
        end }

      min_cf_node.nbrs.each_index{|j|
        min_edge = min_cf_node.edges[j]
        edge = cf_node.edges[j]

        if min_edge > edge
          @min_cf = cf
        end
        if min_edge < edge
          return
        end }}   
  end
end


class CW_BFS
  def initialize(g,r,fc)
    @graph = g
    @root = r
    @first_child = fc
    @q = []
    @parent_pos = []
  end

  def search
    bfi = BFI.new
    @q.push(@root)
    @parent_pos[@root.label] = @root.return_neighbour_pos(@first_child)
    bfi.add_node(@root)
    
    while @q.length != 0
      node = @q.shift
      node.nbrs.each_index {|i|
        nbr_node = node.return_neighbour(@graph,(@parent_pos[node.label]+i)%3)
        edge = node.return_edge(nbr_node)
        if !bfi.has_node(nbr_node)
          bfi.add_node(nbr_node)
          @q.push(nbr_node)
          @parent_pos[nbr_node.label] = nbr_node.return_neighbour_pos(node)
        end
        if !bfi.has_edge(edge)
          bfi.add_edge(edge)
        end}
    end
    return bfi
  end
  
end

class BFI
  attr_accessor :edges, :nodes, :num_nodes, :num_edges
  
  def initialize
    @edges = []
    @nodes = []
    @num_nodes = 0
    @num_edges = 0
  end

  def add_node n
    @nodes[n.label] = @num_nodes
    @num_nodes += 1
  end

  def add_edge e
    @edges[e.abs] = @num_edges
    @num_edges +=1
  end

  def has_node n
    if @nodes[n.label]
      return true
    else
      return false
    end
  end

  def has_edge e
    if @edges[e.abs]
      return true
    else
      return false
    end
  end
    
  
end
    

class Graph
  attr_accessor :adj_list, :num_edges, :num_nodes, :graph_number, :nodes_added, :org_num_nodes, :com_graphs

  def initialize num
    @adj_list = []
    @num_edges = 0
    @org_num_nodes
    @num_nodes = 0
    @graph_number = num
    @com_graphs = 0
  end

  def to_s
    s = "#{@num_nodes}"
    adj_list.each {|n| s += " #{n.num_nbrs} #{n.nbrs[0]} #{n.nbrs[1]} #{n.nbrs[2]}" }
    return s
  end
  
  def read_graph(file,i)
    contents = file[i].split(' ')
    contents.map! { |x| x.to_i }
    @num_nodes = contents[0]
    @org_num_nodes = @num_nodes
    @num_nodes.times {|x| adj_list[x] = Node.new(x,3,"n")}
    for i in 0..(@num_nodes-1)
      for j in 0..2
        adj_list[i].add_neighbour(contents[2+4*i+j])
      end
    end
  end

  def print_graph
    adj_list.each {|x| puts x}
    puts
  end


  def set_edges
    #set all edges to nil
    @adj_list.each { |n| n.edges.map! { |nbr| nil } }
    @num_edges = 0
    
    for i in 0..(@num_nodes-1) #iterate over all the nodes
      for j in 0..(@adj_list[i].num_nbrs - 1) #iterate over all the nbrs of the node
        
        if @adj_list[i].edges[j].nil? 
          node = @adj_list[i]
          nbr_node = node.return_neighbour(self,j)

          #is the node connected to same node mult x?        
          if node.contains_multiple_edges(nbr_node)
            num_me = node.number_of_multiple_edges(nbr_node)

          #walk to the first instance of an edge to nbr_node in node  
          pos = node.return_first_pos(nbr_node)

           #label the edges
          num_me.times{|m| node.edges[(pos+m) % node.num_nbrs] = @num_edges+m}
 
          #repeat for nbr, but opposite order of edge numbers
          pos = nbr_node.return_first_pos(node)

          num_me.times{|m| nbr_node.edges[(pos+m) % nbr_node.num_nbrs] = -((@num_edges-1)+num_me-m)}
          
          @num_edges += num_me
           
          else
            #there are no multiedges. label the edge the edge count and repeat for its neighbour
            index = node.return_neighbour_pos(nbr_node)
            node.edges[index] = @num_edges

            index = nbr_node.return_neighbour_pos(node)
            nbr_node.edges[index] = -@num_edges

            @num_edges += 1
    
          end 
        end       
      end
    end
  end


  def add_node(new_node)
    @adj_list[@num_nodes] = new_node
    @num_nodes += 1
  end

  def bisect
    #bisect each edge with a new c node
    for i in 0..@num_nodes-1
      node = @adj_list[i]
      for j in 0..node.num_nbrs-1
        nbr_node = node.return_neighbour(self,j)
        if nbr_node.type == "n"
          edge_num = node.edges[j]
          new_node = Node.new(@num_nodes,2,"c")
          new_node.add_neighbour(node.label)
          new_node.add_neighbour(nbr_node.label)
          add_node(new_node)
          node.replace_neighbour(nbr_node,new_node,j)
          pos = nbr_node.return_abs_edge_position(edge_num)
          nbr_node.replace_neighbour(node,new_node,pos)
        end
      end
    end
    
    #set new edge numbers
    set_edges
  end

  def add_chords(level)
    require 'debug'
    if level == @num_nodes
      if adj_list[8].nbrs.include? 11
        print_graph
      end
      @com_graphs += 1
      return
    end

    node = @adj_list[level]

    if node.num_nbrs < 3
      candidates = get_candidates node
      candidates.each {|x|
        connect_nodes(x)
        add_chords(level+1)
        disconnect_nodes(x) }
    else
      add_chords(level+1)
    end
  end

  def connect_nodes pos_arr

    node = @adj_list[pos_arr[0]]
    nbr_node = @adj_list[pos_arr[2]]

    node.num_nbrs += 1
    nbr_node.num_nbrs += 1

    
    #order the nbrs of node to insert new nbr in pos
    j = node.num_nbrs-1
    while j > pos_arr[1]
      node.nbrs[j] = node.nbrs[j-1]
      node.edges[j] = node.edges[j-1]
      j -= 1
    end
    node.nbrs[pos_arr[1]] = nbr_node.label
    node.edges[pos_arr[1]] = @num_edges

    j = nbr_node.num_nbrs-1
    while j > pos_arr[3]+1
      nbr_node.nbrs[j] = nbr_node.nbrs[j-1]
      nbr_node.edges[j] = nbr_node.edges[j-1]
      j -= 1
    end
    
    nbr_node.nbrs[pos_arr[3]+1] = node.label
    nbr_node.edges[pos_arr[3]+1] = -@num_edges

    @num_edges += 1    
  end

  def disconnect_nodes pos_arr
    node = @adj_list[pos_arr[0]]
    nbr_node = @adj_list[pos_arr[2]]
    node.remove_node(nbr_node)
    nbr_node.remove_node(node)
  end
  

  #walk around the face getting candidates
  def get_candidates node

    candidates = []

    for i in 0..node.num_nbrs-1
      first_edge = node.edges[i]
      curr_edge = first_edge
      curr_node = node.return_neighbour(self,i)
      pos_to_add_on_node = i
      first = true
      
      while curr_edge != first_edge or first
        first = false
        #if vertex is only degree 2, make it a candidate
        if curr_node != node and curr_node.num_nbrs == 2
          #find the pos of where the edge should be added on curr_node
          pos_to_add_on_nbr = curr_node.edges.index(-curr_edge)
          candidates.push([node.label,pos_to_add_on_node,curr_node.label,pos_to_add_on_nbr])
        end

        #move to the next node
        pos = curr_node.next_pos_after_edge(-curr_edge)
        curr_edge = curr_node.edges[pos]
        curr_node = curr_node.return_neighbour(self,pos)

      end
    end
    return candidates
  end

  def deep_copy
    g = Graph.new(@graph_number)
    g.num_edges = @num_edges
    g.org_num_nodes = @org_num_nodes
    g.num_nodes = @num_nodes
    g.com_graphs = @com_graphs
    @adj_list.each{|x| g.adj_list[x.label] = x.deep_copy}
    return g
  end
  
 end

class Node
  attr_accessor :label, :nbrs, :edges, :num_nbrs, :type

  def initialize(l,n=0,a="")
    @label = l
    @nbrs = []
    @edges = []
    @num_nbrs = n
    @type = a
  end

  def add_neighbour n
    nbrs.push n
  end
  
  def return_first_pos(nbr_node) #return the first instance of a neighbour in a nodes rot. system
    i = 0 
    while (@nbrs[i] == nbr_node.label) and (i < @num_nbrs)
      i += 1
    end
    while (@nbrs[i % @num_nbrs] != nbr_node.label)
      i += 1  
    end
    return (i % @num_nbrs)
  end

  def return_neighbour_pos(nbr_node)
    i = @nbrs.index {|x| x == nbr_node.label}
    if i.nil?
      return nil
    else
      return i
    end
  end

  def return_edge(nbr_node)
    i = return_neighbour_pos(nbr_node)
    return @edges[i]
  end

  def replace_neighbour(old_node,new_node,pos)
    @nbrs[pos] = new_node.label
  end

  def return_abs_edge_position(e)
    return @edges.index {|x| x.abs == e.abs}
  end


  
  def return_neighbour(g,i)
    nbr_label = @nbrs[i]
    return g.adj_list[nbr_label]
  end
  
    
  def to_s
    return "Node: #{label} #{num_nbrs} Neighbours: #{nbrs[0]} (#{edges[0]}), #{nbrs[1]} (#{edges[1]}), #{nbrs[2]} (#{edges[2]})"
  end

  def contains_multiple_edges(nbr_node)
    if number_of_multiple_edges(nbr_node) == 0
      return false
    else
      return true
    end
  end

  def number_of_multiple_edges(nbr_node)
    @nbrs.count(nbr_node.label)
  end

  def next_pos_after_edge edge
    i = (@edges.index {|e| e==edge} + 1) % @num_nbrs
    return i
  end

  def shift_nbrs i
    i.times { nbr = @nbrs.shift
      @nbrs.push nbr}
  end
    

  def remove_node node
    pos = return_neighbour_pos(node)
    @nbrs.delete_at(pos)
    @edges.delete_at(pos)
    @num_nbrs -= 1
  end

  def deep_copy
    n = Node.new(@label,@num_nbrs,@type)
    n.nbrs = [nbrs[0],nbrs[1],nbrs[2]]
    n.edges = [edges[0],edges[1],edges[2]]
    return n
  end

end


class MyFile
  attr_accessor :graphs
  def initialize file_name
    @graphs = IO.readlines(file_name)
  end
end

class Main

  def initialize fn
    @filename = fn
    @total = 0
    @total_nd = 0
  end
  
  def generate_graphs
    f = MyFile.new(@filename)
    f.graphs.each_index{|x|
      g = Graph.new x      
      g.read_graph(f.graphs,x)      
      g.set_edges
      g.bisect
      g.add_chords(g.org_num_nodes)
      @total += g.com_graphs}
  end
end

Main.new("graphs").generate_graphs

File: code/gg.c
/* Disclaimer: only works for graph with multiple edges where edges to the same node are adjacent to each other.
 Code written by David Mitchell based on instruction from Prof Wendy Myrvold University of Victoria */


#include <stdio.h>
#include <stdlib.h>

#define NMAX 512
#define DEG 3

struct Graph
{
    int adj_list[NMAX][DEG]; //a list of the vertices connected to each vertex
    int edge_list[NMAX][DEG]; //a list of the edges which connect vertices
    int num_edges[NMAX]; //the degree of a vertex
    int gap_list[NMAX][DEG]; //the faces associated with each edge
    int total_edges; //the number of edges in a graph
    int total_degree; //the number of degrees
    int num_faces; //the number of faces on a graph
    int num_nodes; //the number of vertices per graph
    int sg_num_nodes; //the number of vertices in the underlying graph
    int graph_number; //the graph number
};

int read_graph(struct Graph *graph, FILE *file); //reads in the graph, returns 0 if the end of file has been reached
void set_edges(struct Graph *graph); //sets the edge numbers
int num_multiedges(int vert, int nbr_vert, struct Graph *graph);
void set_multiedges(int num_me, int edge_count, int vert, int nbr_vert, struct Graph *graph);
void set_gap(struct Graph *graph);
void walk_face(int vert, int nbr_pos, int face_num, struct Graph * graph);
void bisect_graph(struct Graph *graph);
int get_pos(int find_val, int row[], int row_length);
void get_candidates(int v, int *num_candidates,int candidates[NMAX][3], struct Graph *graph);
int common_face(int v1, int v2, struct Graph *graph);
void add_chords(int level, struct Graph *graph, int * com_graphs);
void add_edge(int v1, int v2, int p1, int p2, struct Graph * graph);
void remove_edge (int v1, int v2, struct Graph *graph);

void print_for_cf(struct Graph *graph);
void print_graph(struct Graph *graph);
void print_edge_list(struct Graph *graph);
void print_gap(struct Graph *graph);
void print_faces(struct Graph * graph);

/***************************************************************************************/


int main(int argc, char* argv[])
{
    
    //clear the output streams
    fflush(stdout);
    int com_graphs = 0;
    
    //open the file containing the graph
    FILE *file;
    if (argc > 1)
    {
        file = fopen(argv[1], "r");
    }
    else
    {
        file = stdin;
    }
    
    //read the graph
    struct Graph graph;
    graph.graph_number = 1;
    while (read_graph(&graph,file) != 0)
    {
        set_edges(&graph);
        set_gap(&graph);
        bisect_graph(&graph);
        set_gap(&graph);
        add_chords(graph.sg_num_nodes, &graph, &com_graphs);
        graph.graph_number++;
        //printf("Completed graphs: %d\n",com_graphs);
        com_graphs = 0;
    }
    
}

/***************************************************************************************/

int read_graph(struct Graph *graph, FILE *file)

{
    int i;
    int j;
    
    if (!fscanf(file,"%d",&graph->num_nodes))
    {
        fprintf(stderr, "error reading in graph (error using scanf)\n");
        exit(1);
    }
    
    //if we've reached the end of the file return 0
    if (feof(file))
    {
        return 0;
    }
    
    //check the total number of nodes
    if (graph->num_nodes < 1)
    {
        fprintf(stderr, "error reading in graph\n");
        exit(1);
    }
    
    graph->total_degree = 0;
    
    // for each row use scanf to read the # of edges and insert them in the matrix
    for (i = 0; i < graph->num_nodes; i++)
    {
        if (!fscanf(file,"%d",&graph->num_edges[i]))
        {
            fprintf(stderr, "error reading in graph (error using scanf)\n");
            exit(1);
        }
        
        // check the edge numbers arent incorrect
        if (graph->num_edges[i] != DEG)
        {
            fprintf(stderr, "error reading in graph (graph must be %2d regular)\n",DEG);
            exit(1);
        }
        
        graph->total_degree += graph->num_edges[i];
        
        //read num_edges[i] nodes into the graph
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            if (!fscanf(file,"%d",&graph->adj_list[i][j]))
            {
                fprintf(stderr, "error reading in graph (error using scanf)\n");
                exit(1);
            }
            
            //input clearly can't exceed the number of nodes and can't be less than 0
            //as well, the graph contains no self loops
            if (graph->adj_list[i][j] > graph->num_nodes || graph->adj_list[i][j] < 0 || graph->adj_list[i][j] == i)
            {
                fprintf(stderr, "error reading in graph (incorrect node)\n");
                exit(1);
            }
        }
    }
    
    graph->total_edges = graph->total_degree/2;
    graph->num_faces = 2 + graph->total_edges - graph->num_nodes;
    graph->sg_num_nodes = graph->num_nodes;
    
    return 1;
}


/***************************************************************************************/

void set_edges(struct Graph *graph)

{
    int i;
    int j;
    int k;
    int num_me;
    int nbr_vert;
    int edge_count;
    
    //set all the edges to -1
    for (i = 0; i < graph->num_nodes; i++)
    {
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            graph->edge_list[i][j] = -1;
        }
    }
    
    //set all the edges
    edge_count = 2;
    for (i = 0; i < graph->num_nodes; i++)
    {
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            if (graph->edge_list[i][j] == -1) //the edge has not been set
            {
                //determine if there are multiple edges to the same vertex
                nbr_vert = graph->adj_list[i][j];
                num_me = num_multiedges(i,nbr_vert,graph);
                if (num_me > 1)
                {
                    set_multiedges(num_me,edge_count,i,nbr_vert,graph);
                    edge_count = edge_count+num_me;
                }
                else
                {
                    graph->edge_list[i][j] = edge_count;
                    
                    //change the edge on the home row of the neighbour vertex after finding it
                    k = 0;
                    while (graph->adj_list[nbr_vert][k] != i)
                    {
                        k++;
                    }
                    graph->edge_list[nbr_vert][k] = -edge_count;
                    edge_count++;
                }
            }
            
        }
    }
}

/***************************************************************************************/

void set_multiedges(int num_me, int edge_count, int vert, int nbr_vert, struct Graph * graph)

{
    int i;
    int j;
    
    //walk to the first instance of an edge to nbr_vert on the home row of vert
    i = 0;
    while (graph->adj_list[vert][i] == nbr_vert && i < graph->num_edges[vert])
    {
        i++;
    }
    while (graph->adj_list[vert][i % graph->num_edges[vert]] != nbr_vert)
    {
        i++;
    }
    
    //now label the edges
    for (j = 0; j < num_me; j++)
    {
        graph->edge_list[vert][(i + j) % graph->num_edges[vert]] = edge_count+j;
    }
    
    //repeat for the nbr vertex, but opposite order of edge numbers
    i = 0;
    while (graph->adj_list[nbr_vert][i] == vert && i < graph->num_edges[nbr_vert])
    {
        i++;
    }
    while (graph->adj_list[nbr_vert][i % graph->num_edges[nbr_vert]] != vert)
    {
        i++;
    }
    
    for (j = 0; j < num_me; j++)
    {
        graph->edge_list[nbr_vert][(i + j) % graph->num_edges[vert]] = -((edge_count-1)+num_me-j);
    }
    
}
/***************************************************************************************/


int num_multiedges(int vert, int nbr_vert, struct Graph *graph)

{
    int i;
    int count = 0;
    
    for (i = 0; i < graph->num_edges[vert]; i++)
    {
        if (graph->adj_list[vert][i] == nbr_vert)
        {
            count++;
        }
    }
    return count;
}


/***************************************************************************************/

void set_gap(struct Graph * graph)

{
    int i;
    int j;
    int face_num;
    
    //initialize the gap matrix to -1
    for (i = 0; i < graph->num_nodes; i++)
    {
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            graph->gap_list[i][j] = -1;
        }
    }
    
    face_num = 0;
    for (i = 0; i < graph->num_nodes; i++)
    {
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            //if the edge is not part of a face, walk the face and increment number of faces
            if (graph->gap_list[i][j] == -1)
            {
                walk_face(i, j, face_num, graph);
                face_num++;
            }
        }
    }
    
}

/***************************************************************************************/
void walk_face (int vert, int nbr_pos, int face_num, struct Graph *graph)

{
    int i;
    int first_edge = graph->edge_list[vert][nbr_pos];
    int next_edge = -1;
    int next_vert = -1;
    int curr_edge = first_edge;
    int curr_vert = graph->adj_list[vert][nbr_pos];
    
    while (next_edge != first_edge)
    {
        //move to the position following the last edge
        i = 0;
        while(graph->edge_list[curr_vert][i] != -curr_edge)
        {
            i++;
        }
        i++;
        
        //set the edge following the previous to be part of the face
        graph->gap_list[curr_vert][i % graph->num_edges[curr_vert]] = face_num;
        next_edge = graph->edge_list[curr_vert][i % graph->num_edges[curr_vert]];
        next_vert = graph->adj_list[curr_vert][i % graph->num_edges[curr_vert]];
        
        curr_edge = next_edge;
        curr_vert = next_vert;
    }
}

/***************************************************************************************/

void bisect_graph(struct Graph *graph)

{
    int i;
    int j;
    int nbr_vert;
    int new_vert;
    int edge_num;
    int pos;
    
    for (i = 0; i < graph->sg_num_nodes; i++)
    {
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            //if a vertex neighbour num is less than the num of vertices, it hasn't been bisected
            if (graph->adj_list[i][j] < graph->sg_num_nodes)
            {
                nbr_vert = graph->adj_list[i][j];
                edge_num = graph->edge_list[i][j];
                pos = get_pos(-edge_num, graph->edge_list[nbr_vert], graph->num_edges[nbr_vert]);
                
                new_vert = graph->num_nodes;
                graph->num_edges[new_vert] = 0;
                
                //set the home row of the vis vert to point to new vert
                graph->adj_list[i][j] = new_vert;
                
                //set the new vert to point to vis vert
                graph->adj_list[new_vert][0] = i;
                graph->total_edges++;
                graph->num_edges[new_vert]++;
                
                //set the home row of the nbr vert to point to new vert
                graph->adj_list[nbr_vert][pos] = new_vert;
                
                //set the new vert to point to nbr vert
                graph->adj_list[new_vert][1] = nbr_vert;
                graph->num_edges[new_vert]++;

                
                graph->num_nodes++;
            }
        }
    }
    set_edges(graph);
}

/***************************************************************************************/

void add_chords(int level, struct Graph *graph, int * com_graphs)
{
    int i;
    int v;
    int num_candidates;
    int candidates[NMAX][3];
    
    v = level;
    if (v == graph->num_nodes)
    {
        print_for_cf(graph);
        *com_graphs = *com_graphs + 1;
        return;
    }
    
    if (graph->num_edges[v] < DEG)
    {
        get_candidates(v, &num_candidates, candidates, graph);
        for (i = 0; i < num_candidates; i++)
        {
            add_edge(v, candidates[i][0],candidates[i][1],candidates[i][2], graph);
            add_chords(v+1, graph,com_graphs);
            remove_edge(v, candidates[i][0], graph);
        }
    }
    else
    {
        add_chords(v+1, graph, com_graphs);
    }
    
}

/***************************************************************************************/
void get_candidates(int v, int *num_candidates,int candidates[NMAX][3], struct Graph *graph) //
{
    
    
    /*hv (home vertex) pos and tv (to vertex) pos keep track of the positions where the new edge is to be added
    for the home vertex, the edge is added 1 before the vertex it initially leads to.
    for the to vertex, the edge is added 1 after the edge leading to it  */
    
    int i, j, k, c, hv_pos;
    c = 0;
    
    *num_candidates = 0;
    
    for (j = 0; j < DEG-1; j++)
    {
        int first_edge = graph->edge_list[v][j];
        int next_edge = -1;
        int next_vert = -1;
        int curr_edge = first_edge;
        int curr_vert = graph->adj_list[v][j];
        hv_pos = j;
        
        while (next_edge != first_edge)
        {
            //move to the position following the last edge
            i = 0;
            while(graph->edge_list[curr_vert][i] != -curr_edge)
            {
                i++;
            }
            i++;
            
            //if the vertex is only degree 2, make it a candidate
            if (graph->num_edges[curr_vert] == 2 && curr_vert != v)
            {
                candidates[c][0] = curr_vert;
                candidates[c][1] = hv_pos;
                
                //find the position of where the edge should be added
                k = 0;
                while (graph->edge_list[curr_vert][k] != -curr_edge)
                {
                    k++;
                }
                
                candidates[c][2] = k;
                c++;
                *num_candidates = *num_candidates + 1;
            }
            
            
            next_edge = graph->edge_list[curr_vert][i % graph->num_edges[curr_vert]];
            next_vert = graph->adj_list[curr_vert][i % graph->num_edges[curr_vert]];
            
            curr_edge = next_edge;
            curr_vert = next_vert;
        }
    }
    
}

/***************************************************************************************/
void add_edge(int v1, int v2, int p1, int p2, struct Graph * graph)

{
    int j;
    
    //increase the num of edges of v1 and v2
    graph->num_edges[v1]++;
    graph->num_edges[v2]++;
    
    //find the position for the new edge on v1
    
    for (j = graph->num_edges[v1] - 1; j > p1; j--)
    {
        graph->adj_list[v1][j] = graph->adj_list[v1][j-1];
        graph->edge_list[v1][j] = graph->edge_list[v1][j-1];
    }
    
    graph->adj_list[v1][p1] = v2;
    graph->edge_list[v1][p1] = graph->total_edges+2;
    
    //find the position for the new edge on v2
    
    for (j = graph->num_edges[v2] - 1; j > p2+1; j--)
    {
        graph->adj_list[v2][j] = graph->adj_list[v2][j-1];
        graph->edge_list[v2][j] = graph->edge_list[v2][j-1];
    }
    
    graph->adj_list[v2][p2+1] = v1;
    graph->edge_list[v2][p2+1] = -(graph->total_edges+2);
    
    graph->total_edges++;
    graph->num_faces++;
    
    set_gap(graph);
    
}

/***************************************************************************************/

void remove_edge (int v1, int v2, struct Graph *graph)

{
    int i;
    int j;
    int k;
    
    //find the position of v2
    i = 0;
    while (graph->adj_list[v1][i] != v2)
    {
        i++;
    }
    
    //copy the array but bypass pos of v2
    k = 0;
    for (j = 0; j < graph->num_edges[v1]; j++)
    {
        if (j != i)
        {
            graph->adj_list[v1][k] = graph->adj_list[v1][j];
            graph->edge_list[v1][k] = graph->edge_list[v1][j];
            k++;
        }
    }
    graph->num_edges[v1]--;
    
    //do the same for v2
    i = 0;
    while (graph->adj_list[v2][i] != v1)
    {
        i++;
    }
    
    //copy the array but bypass pos of v1
    k = 0;
    for (j = 0; j < graph->num_edges[v2]; j++)
    {
        if (j != i)
        {
            graph->adj_list[v2][k] = graph->adj_list[v2][j];
            graph->edge_list[v2][k] = graph->edge_list[v2][j];
            k++;
        }
    }
    graph->num_edges[v2]--;
    
    graph->total_edges--;
    graph->num_faces--;
    
    set_gap(graph);
}

/***************************************************************************************/

int common_face(int v1, int v2, struct Graph *graph)

{
    int i;
    int j;
    for (i = 0; i < graph->num_edges[v1]; i++)
    {
        for (j = 0; j < graph->num_edges[v2]; j++)
        {
            if (graph->gap_list[v1][i] == graph->gap_list[v2][j])
            {
                return graph->gap_list[v1][i];
            }
            
        }
    }
    return -1;
}


/***************************************************************************************/

int get_pos(int find_val, int row[], int row_length)

{
    int i;
    for (i = 0; i < row_length; i++)
    {
        if (row[i] == find_val)
        {
            return i;
        }
    }
    return -1;
}
/***************************************************************************************/

void print_graph(struct Graph *graph)

{
    int i;
    int j;
    
    printf("%d ",graph->num_nodes);
    for (i = 0; i < graph->num_nodes; i++)
    {
        printf("%d ", graph->num_edges[i]);
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            printf("%d ", graph->adj_list[i][j]);
        }
    }
    printf("\n");
}

/***************************************************************************************/

void print_for_cf(struct Graph *graph)

{
    int i;
    int j;
    printf("%d ", graph->graph_number);
    printf("%d ", graph->num_nodes);
    for (i = 0; i < graph->num_nodes; i++)
    {
        printf ("%d ", graph->num_edges[i]);
        
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            printf("%d %d ", graph->adj_list[i][j], graph->edge_list[i][j]);
        }
        
    }
    printf("\n");
}




/***************************************************************************************/

void print_edge_list(struct Graph *graph)

{
    printf("Printing edge list\n");
    int i;
    int j;
    for (i = 0; i < graph->num_nodes; i++)
    {
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            printf("%d ", graph->edge_list[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

/***************************************************************************************/

void print_gap(struct Graph *graph)

{
    printf("Printing gap list\n");
    int i;
    int j;
    for (i = 0; i < graph->num_nodes; i++)
    {
        for (j = 0; j < graph->num_edges[i]; j++)
        {
            printf("%d ", graph->gap_list[i][j]);
        }
        printf("\n");
    }
    printf("\n");
}

/***************************************************************************************/

void print_faces(struct Graph * graph)

{
    int i;
    int j;
    int k;
    int l;
    int first_edge;
    int next_edge;
    int next_vert;
    int curr_edge;
    int curr_vert;
    
    printf("Printing faces\n");
    for (i = 0; i < graph->num_faces; i++)
    {
        printf("Printing face #%d:\n", i);
        
        j = 0;
        k = 0;
        while (graph->gap_list[j][k] != i)
        {
            if (k < graph->num_edges[j])
            {
                k++;
            }
            else
            {
                j++;
                k = 0;
            }
        }
        
        first_edge = graph->edge_list[j][k];
        next_edge = -1;
        next_vert = -1;
        curr_edge = first_edge;
        curr_vert = graph->adj_list[j][k];
        
        while (next_edge != first_edge)
        {
            //move to the position following the last edge
            l = 0;
            while(graph->edge_list[curr_vert][l] != -curr_edge)
            {
                l++;
            }
            l++;
            
            next_edge = graph->edge_list[curr_vert][l % graph->num_edges[curr_vert]];
            next_vert = graph->adj_list[curr_vert][l % graph->num_edges[curr_vert]];
            
            printf("Vertex %d to vertex %d edge number %d.\n", curr_vert, next_vert, curr_edge);
            
            curr_edge = next_edge;
            curr_vert = next_vert;
        }
        printf("\n\n");
    }
    printf("\n");
}


